File: go-backend-interview-prep/02_data_structures/08_json_basics/main.go
package main

import (
	"encoding/json"
	"fmt"
)

// Struct tags control JSON field names and behavior.
type User struct {
	Name     string `json:"name"`
	Email    string `json:"email,omitempty"` // omit if empty string
	Age      int    `json:"age"`
	Password string `json:"-"` // never include in JSON
	Active   bool   `json:"active,omitempty"`
}

func main() {
	// === MARSHAL: struct -> JSON ===
	fmt.Println("--- Marshal ---")
	u := User{
		Name:     "Alice",
		Email:    "alice@example.com",
		Age:      30,
		Password: "secret123",
		Active:   true,
	}

	data, err := json.Marshal(u)
	if err != nil {
		fmt.Println("marshal error:", err)
		return
	}
	fmt.Println("JSON:", string(data))
	// Password is excluded (-), all other fields present

	// === MARSHAL with omitempty ===
	fmt.Println("\n--- omitempty ---")
	u2 := User{Name: "Bob", Age: 25}
	// Email is "" -> omitted; Active is false -> omitted
	data2, _ := json.Marshal(u2)
	fmt.Println("JSON:", string(data2))

	// === PRETTY PRINT ===
	fmt.Println("\n--- MarshalIndent ---")
	pretty, _ := json.MarshalIndent(u, "", "  ")
	fmt.Println(string(pretty))

	// === UNMARSHAL: JSON -> struct ===
	fmt.Println("--- Unmarshal ---")
	input := `{"name":"Charlie","email":"c@test.com","age":28,"active":true}`
	var u3 User
	err = json.Unmarshal([]byte(input), &u3) // must pass pointer
	if err != nil {
		fmt.Println("unmarshal error:", err)
		return
	}
	fmt.Printf("parsed: %+v\n", u3)

	// === Unknown fields are silently ignored ===
	fmt.Println("\n--- Unknown fields ---")
	weird := `{"name":"Dave","unknown_field":"ignored","age":35}`
	var u4 User
	json.Unmarshal([]byte(weird), &u4)
	fmt.Printf("unknown fields ignored: %+v\n", u4)

	// === Partial JSON (missing fields get zero values) ===
	fmt.Println("\n--- Partial JSON ---")
	partial := `{"name":"Eve"}`
	var u5 User
	json.Unmarshal([]byte(partial), &u5)
	fmt.Printf("partial: %+v\n", u5) // Age=0, Email=""

	// === Marshal a slice ===
	fmt.Println("\n--- Slice of structs ---")
	team := []User{
		{Name: "Alice", Age: 30},
		{Name: "Bob", Age: 25},
	}
	teamJSON, _ := json.Marshal(team)
	fmt.Println("team:", string(teamJSON))

	// === Marshal a map ===
	fmt.Println("\n--- Map ---")
	config := map[string]string{
		"host": "localhost",
		"port": "8080",
	}
	cfgJSON, _ := json.Marshal(config)
	fmt.Println("config:", string(cfgJSON))
}

File: go-backend-interview-prep/02_data_structures/08_json_basics/main.py
# Python equivalent of JSON Basics (compare with main.go)
# Go encoding/json -> Python json module
# Go struct tags -> Python dict keys or dataclass + custom encoder

import json
from dataclasses import dataclass, asdict


@dataclass
class User:
    name: str = ""
    email: str = ""
    age: int = 0
    password: str = ""   # excluded from JSON (like Go's json:"-")
    active: bool = False


def user_to_dict(u: User, omit_empty: bool = False) -> dict:
    """Mimics Go struct tags: exclude password, optionally omit zero values."""
    d = {"name": u.name, "age": u.age}

    if not omit_empty or u.email:
        d["email"] = u.email
    if not omit_empty or u.active:
        d["active"] = u.active
    # password always excluded (like json:"-")
    return d


def main():
    # === MARSHAL: object -> JSON string ===
    print("--- Marshal ---")
    u = User(name="Alice", email="alice@example.com", age=30,
             password="secret123", active=True)

    data = json.dumps(user_to_dict(u))
    print("JSON:", data)

    # === MARSHAL with omitempty ===
    print("\n--- omitempty ---")
    u2 = User(name="Bob", age=25)
    data2 = json.dumps(user_to_dict(u2, omit_empty=True))
    print("JSON:", data2)

    # === PRETTY PRINT ===
    print("\n--- Pretty print ---")
    pretty = json.dumps(user_to_dict(u), indent=2)
    print(pretty)

    # === UNMARSHAL: JSON string -> dict -> object ===
    print("--- Unmarshal ---")
    input_json = '{"name":"Charlie","email":"c@test.com","age":28,"active":true}'
    d = json.loads(input_json)
    u3 = User(**{k: v for k, v in d.items() if k in User.__dataclass_fields__})
    print(f"parsed: {u3}")

    # === Unknown fields silently ignored ===
    print("\n--- Unknown fields ---")
    weird = '{"name":"Dave","unknown_field":"ignored","age":35}'
    d = json.loads(weird)
    u4 = User(name=d.get("name", ""), age=d.get("age", 0))
    print(f"unknown fields ignored: {u4}")

    # === Partial JSON ===
    print("\n--- Partial JSON ---")
    partial = '{"name":"Eve"}'
    d = json.loads(partial)
    u5 = User(name=d.get("name", ""), age=d.get("age", 0))
    print(f"partial: {u5}")

    # === List of dicts ===
    print("\n--- List ---")
    team = [{"name": "Alice", "age": 30}, {"name": "Bob", "age": 25}]
    print("team:", json.dumps(team))

    # === Dict -> JSON ===
    print("\n--- Dict ---")
    config = {"host": "localhost", "port": "8080"}
    print("config:", json.dumps(config))


if __name__ == "__main__":
    main()
